### ISSUE TITLE
Recent change in maybe_allowed is incompatible.

### ISSUE THREAD

**Issue by mmzeeman** [zotonic/zotonic#1398]
A recent change in z_acl:maybe_allowed is incompatible.

In a recent commit https://github.com/zotonic/zotonic/commit/490af340a3b85da1c2ed4f5dc7715f328931f768#diff-1409e9b8609a6090811c9ef6eaef80cdR83 a subtle change has been made.

Function `z_acl:maybe_allowed/3` now resolves atom object values with `m_rsc:rid/2`. Sometimes these atom values are not resources. Like the use check for which an exception has been made. For channel we have acl checks for live processes like chat-request fsm's and cobrowse sessions which are not resources. Checks for are made by custom acl modules with a higher prio. These checks now fail because atom values objects are replaced with undefined by the call to m_rsc:rid.

Is this something we want? Do we always want atom values in object to be resources and use rid to resolve them? I can change my custom acl checks by using a tuple instead of an atom. Or should the call to rid by done later in the chain? If that is the case we should change the `use` acl check too and use a tuple value to check if somebody can use a module.

Ping @mworrell @ddeboer @arjan

**Comment by ddeboer** [comment]
> Is this something we want? Do we always want atom values in object to be resources and use rid to resolve them?

No, I don’t think we do. We can consider https://github.com/zotonic/zotonic/commit/490af340a3b85da1c2ed4f5dc7715f328931f768#diff-1409e9b8609a6090811c9ef6eaef80cdR83 a mistake.

I agree with:

> maybe it is better to pass the original value when m_rsc:rid/2 returns undefined. That would be backward compatible.

**Comment by mmzeeman** [comment]
After giving it some thought, passing the original value after rid has returned undefined would give some side effects when resources accidentally have been made with that name. So that is probably not what we want to do either.

**Comment by ddeboer** [comment]
Good point. So do we solve this so that all of the following work without side-effects?

1. `z_acl:is_allowed(view, my_rsc_unique_name,  C)` 
2. `z_acl:is_allowed(view, 123,  C)` (a resource id)
3. `z_acl:is_allowed(view, my_custom_non_resource_name_atom, C)`
4. `z_acl:is_allowed(view, 987, C)` (some custom integer, not a resource id)

It seems we can’t differentiate properly between 1/3 and 2/4. I see two options

* We assume that for the typical resource actions `view|update|insert` the second argument is a resource id or name. For all other actions (`my_custom_permission`) we don’t assume this.
* Or we just remove the `rid` here, but that will cause 1 above not to work any longer.

**Comment by mmzeeman** [comment]
The best solution is probably that we tag the object values? 

```
z_acl:is_allowed(view, {rsc, 123}, C)
z_acl:is_allowed(view, {rsc, my_rsc_unique_name}, C)
z_acl:is_allowed(use, {module, mod_admin}, C)
```

But that is not backward compatible of course.

I think it is better to remove the rid call for now. The z_acl currently also needs knowledge of the underlying acl modules to function because it has to filter for the `use` action. 

BTW. The user group acl check also contains checks for atom object values.

```
203 acl_is_allowed(#acl_is_allowed{action=insert, object=Cat}, Context) when is_atom(Cat) ->
204     can_insert(Cat, Context);
```

With this change Cat can't be an atom anymore. So I'm not entirely sure if this change has other side effects.

**Comment by ddeboer** [comment]
Either tags, or records, as we also [find in `acl_user_group_checks`](https://github.com/zotonic/zotonic/blob/master/modules/mod_acl_user_groups/support/acl_user_groups_checks.erl#L201), such as `#acl_rsc{}` and `#acl_media{}`.

**Comment by mmzeeman** [comment]
That sounds like a good idea to do in the master branch. It will be incompatible with the stable branch anyhow.

**Comment by mworrell** [comment]
An `#acl_rsc{}` could be a good solution.
The *z_acl* routines could map integers to `#acl_rsc{}` records, maintaining the notion that an integer is a resource id, unless proven otherwise.

**Comment by ddeboer** [comment]
@mmzeeman Fixed in #1400. Please review.

**Comment by mmzeeman** [comment]
Yep... I can't run our site on master because of the latest cowboy merge, so I tested it by manually removing the lines from the stable branch. Works. :+1:
